### ISSUE TITLE
Feature Request : Prefix returned errors with something such as `Error:`

### ISSUE THREAD

**Issue by chilts** [magefile/mage#27]
In the video demo your example of `return errors.New("boo!")` just prints out "boo!" and whilst the `echo $?` shows a non-zero exit code, there is nothing on the screen to visually show the command failed.

I know some people have prompts which display something when the previous command failed, however I think it would be nice to show it visually too. Nothing too big, but just big enough.

For example:

```
$ mage install
Error: boo!
```

As a comparison let's just look at what `make` does. We don't have to copy this nor am I holding it up as the ideal, but just a curiosity. For example, if my target just tried to cat a non-existant file:

```
$ make target
cat hello.go
cat: hello.go: No such file or directory
Makefile:2: recipe for target 'target' failed
make: *** [build] Error 1
```

Which (I think) is made up of the command, stderr from the command and then output from make itself. I'm happy to keep it real simple as above and just prefix the output with `Error:`.

(Note: I think we could or should also do the same for functions that don't return errors, so I'm not sure how we'd do that - perhaps if they wrote something on stderr? This comment is a non-goal of this ticket and perhaps another ticket should be opened, however it's pertinent to raise this here in case it impacts what we do here.)

**Comment by chilts** [comment]
Or perhaps functions that don't return errors are designed to never fail ... in which case, we don't need to do anything else. :)

**Comment by natefinch** [comment]
functions that don't have an error return can only fail if one of their dependencies fail.

I just pushed a PR that includes a bunch of command execution helpers that default to writing stderr from the command to stderr of mage, and cause mage to exit with the exit code the target command failed with.  I'll do some testing on the error output and make sure it's obvious when things fail.

But in the simple case where you have a `func Foo()` error target, it absolutely should prepend Error: to the output to make it obvious the thing failed.

**Comment by chilts** [comment]
Excellent. I'll give that a go - thanks. I might wait until your branch #26 is merged and pushed.

**Comment by natefinch** [comment]
Yeah, it's kind of a beast. I should really split it up.

**Comment by chilts** [comment]
Alrighty, I think I need your help @natefinch :)

I got this far (after a wild goose chase in `mage/main.go`) I figured that it must somehow compile my own `package main` with something it runs, hence I found the template. Here is pretty much all I changed.

```
$ git staged
diff --git a/mage/template.go b/mage/template.go
index 3815fdd..adca5f8 100644
--- a/mage/template.go
+++ b/mage/template.go
@@ -79,7 +79,7 @@ func main() {
        case "{{lower .Name}}":
                {{if .IsError -}}
                        if err := {{.Name}}(); err != nil {
-                               fmt.Println(err)
+                               fmt.Printf("Error: %s\n", err)
                                os.Exit(1)
                        }
                {{else -}}
```

Now I'm looking at adding a test. However, from what I can see all of the tests that currently exist that return an error return `nil`, so when I run `go test -race ./...` or just `go test ./...`, neither of them fail even though I expected it to, which is what roused my suspicion.

I was going to add something to `mage/mage/main_test.go` but I think I need to add something to `mage/mg/errors.go` and `mage/mg/errors_test.go` though currently I **think** I can see it is testing the exist status, but not the output (which it **does** seem to do in `mage/mage/main_test.go` in `TestGoRun`). Help appreciated. :)

**Comment by natefinch** [comment]
Actually that's just about all you need.  That and line 42, where there's a defer that catches panics and prints them.  Add "Error:", to both print statements, and I think that's all you need.

**Comment by natefinch** [comment]
You shouldn't need to touch mg at all. All it does is produce errors with an ExitStatus that the generated mainfile understands.

There's no output test, so it's definitely good to make one.  you can have a test that runs a target that returns an `errors.New("Foobar")` and check that the stdout is "Error: Foobar".

**Comment by chilts** [comment]
Awesome. Pull request sent. I managed to figure out tests for the output for both a regular returned error and I added two tests for a function and a error function that both panic.

I'm happy to refactor a little bit in those tests so it is table driven rather than having three functions that are essentially checking the same thing. Does that sound okay or are you happy to merge and then I can convert it to being table driven?
